editors/code/src/installation/download_file.ts

```diff
@@ -27,8 +27,7 @@ export async function downloadFile(
             readBytes += chunk.length;
             onProgress(readBytes, totalBytes);
         })
-        .on("end", resolve)
         .on("error", reject)
-        .pipe(fs.createWriteStream(destFilePath))
+        .pipe(fs.createWriteStream(destFilePath).on("close", resolve))
```

Comment:
It's recommended to use [`stream.pipeline`](https://nodejs.org/api/stream.html#stream_stream_pipeline_streams_callback) in all cases over `stream.pipe`. Formerly this functionality was only available in the [`pump`](https://github.com/mafintosh/pump) and `pumpify` packages, but was added two years ago to fix the problems around `pipe` (in particular: error handling, and correctly detecting close events). Also when used with `util.promisify` it allows simplifying Promise construction for the pipeline:

```rust
const pipeline = util.promisify(stream.pipeline);

let bs = response.body;
  .on("data", (chunk: Buffer) => {
    readBytes += chunk.length;
    onProgress(readBytes, totalBytes);
  });
await pipeline(bs, fs.createWriteStream(destFilePath);
```
_nb:_ I used to be on the node streams wg up until a few years ago 😅

Reply:
@yoshuawuyts , thank you for the suggestion, I did look into `stream.pipeline()`, but then reverted back, since it doesn't resolve the promise when the write-stream is closed. You can see my comments about truing trying to incorporate it in this MR, but still it didn't fix the error, but my initial approach with `.on("close")` did, anyway... Maybe this is a bug in NodeJS?

Reply:
Though, no, I cannot reproduce that again with `stream.pipeline()`... I need to recheck it on my home laptop where I did saw it was not working. Or maybe I just was overwhelmed by the bug and somehow was running the wrong code? I am not sure... Thank you for the pointer anyway, I'll double-recheck that `stream.pipeline()` does resolve the promise when the write-stream is closed today!

Reply:
Though, no, here is a [gist of the reproduction](https://gist.github.com/Veetaha/21c6bcab38cde97445364eaa9c0ff89b).
When I run it, I get `RESOLVED` printed before `CLOSED`, though there is no `EBUSY` error on my current PC...
![image](https://user-images.githubusercontent.com/36276403/74334379-77e8af00-4da2-11ea-80b7-86f32aeae5d1.png)

Reply:
Okay, so this turned out to be a fun one.

In the example you're listening for the `"close"` event, which you would rightfully assume is called "at the end of the pipeline". However that's not what it does: it's emitted when the underlying file descriptor is closed. Which because Node.js doesn't have destructors it's emitted when the program is closed: which means it's last.

However if you listen for the `"finish"` event instead, it will trigger when you expect it to: namely when the writing to the writer has finished. The reason this doesn't also close the file descriptor is because it's possible to reuse the same write stream in multiple pipelines and the descriptor should be kept open.

If you want to close the file descriptor after the pipeline has finished, call `.destroy()` on the stream. The docs don't mention this, but it's the standard way of doing it and [I checked the source to double check](https://github.com/nodejs/node/blob/master/lib/internal/fs/streams.js#L415).

Either way: this is definitely harder than it should be, but at least with `pipeline` it's harder to get wrong.

## Example

I believe the downloadFile function should be defined like this to correctly wait for it to finish, and then close the fd:
```js
async function downloadFile(url, destFilePath) {
    let source = await fetch(url).body;
    let sink = fs.createWriteStream(destFilePath, { mode: 0o755 });
    await pipeline(source, sink);
    sink.destroy()
}
```

Reply:
Thank you for the explanations. I did recheck that the first vairant without explicit `.destroy()` still yields `EBUSY`

Reply:
But the second variant yields `EBUSY` too...

Reply:
Here is the output on my home laptop. [Here is the code I ran and output I got](https://gist.github.com/Veetaha/37dcb5ca433ef319dcdc369b345aaece)

Reply:
I am unsure, but if the issue is that the fd is not yet closed at the end of that code snippit, you may also want to promisify `sink.destroy()`.

Reply:
@matklad, I did double-rechecked the pipeline API, even with an explicit `destroy()` call and it still doesn't work on my laptop on Windows. 
Tried to call `.end()` ([which moreover ensures that the stream is flushed](https://nodejs.org/docs/latest-v12.x/api/stream.html#stream_writable_end_chunk_encoding_callback)) instead of `.destroy()`, but that didn't change the day.
I get `EBUSY` error as originally. I do wish we could use this API and tried to switch to it while fixing the original bug, but anyway, it somehow doesn't work properly.
I'll create an issue under nodejs repo and if we are doing something wrong they should help, otherwise I propose to leave explicit `pipe()` calls and to merge #3092, since it does work now. Once we figure out what is going one I'll gladly switch to `stream.pipeline()`

Reply:
We cannot promisify neither `.destroy()`, neither `.end()`, they are synchronous and don't accept any callbacks. One thing we can do is do the same 
```ts
const destFileStream = fs.createWriteStream(path, { mode });
await stream.pipeline(response.body, destFileStream);
stream.end();
return new Promise((resolve) => destFileStream.on("close", resolve));
```
This looks quite hacky...

Reply:
Hmm, wait, `stream.end()` does accept an optional callback...

Reply:
No `stream.end()` accepts a callback that is just attached to `'finish'` event...

I suppose we can use the snippet I wrote higher for now...

Reply:
I've switched to `stream.pipeline()` with `.on("close")` in #3116. Let's leave it like that for now...
@matklad ?

Reply:
FYI: I have created an issue on `nodejs` repo regarding `stream.destroy()` not releasing the file handle immediatly.
https://github.com/nodejs/node/issues/31776